### PR TITLE
CommonClient: implement check_locations to send missing locations only

### PR DESCRIPTION
**Pull request by Berserker66**
## What is this fixing or adding?
check_locations, a method that filters location checks before potentially bothering to send stuff over the wire

## How was this tested?
once with Factorio.

Also included is LttP which hopefully someone tests at some point, though it should be fairly safe if Factorio worked.

### CODE REVIEW

File: CommonClient.py [hunk 1]
```diff
@@ -459,6 +459,14 @@ async def send_connect(self, **kwargs: typing.Any) -> None:
         await self.send_msgs([payload])
         await self.send_msgs([{"cmd": "Get", "keys": ["_read_race_mode"]}])
 
+    async def check_locations(self, locations: typing.Collection[int]) -> set[int]:
+        """Send new location checks to the server. Returns set of actually new locations that were send."""
+        locations = set(locations)
+        locations.update(self.missing_locations)
```
Comment (qwint) [inline CommonClient.py:465]:
```suggestion
        locations &= self.missing_locations
```
or i guess set.intersection_update() ?
set.update() is for union not intersection

**Review by qwint — APPROVED** [review]
pure code review for now, I have a commonclient game so I may come back and test further later

File: CommonClient.py [hunk 2]
```diff
@@ -459,6 +459,13 @@ async def send_connect(self, **kwargs: typing.Any) -> None:
         await self.send_msgs([payload])
         await self.send_msgs([{"cmd": "Get", "keys": ["_read_race_mode"]}])
 
+    async def check_locations(self, locations: typing.Collection[int]) -> set[int]:
+        """Send new location checks to the server. Returns set of actually new locations that were send."""
```
Comment (ScipioWright) [inline CommonClient.py:463]:
```suggestion
        """Send new location checks to the server. Returns the set of actually new locations that were sent."""
```

File: CommonClient.py [hunk 3]
```diff
@@ -459,6 +459,13 @@ async def send_connect(self, **kwargs: typing.Any) -> None:
         await self.send_msgs([payload])
         await self.send_msgs([{"cmd": "Get", "keys": ["_read_race_mode"]}])
 
+    async def check_locations(self, locations: typing.Collection[int]) -> set[int]:
+        """Send new location checks to the server. Returns the set of actually new locations that were sent."""
+        locations = set(locations) & self.missing_locations
```
Comment (agilbert1412) [inline CommonClient.py:464]: Any reason why you aren't asking for `locations` to be a Set in the first place? Factorio's is a set, and from what I can see, alttp could easily be a set as well

**Review by agilbert1412 — APPROVED** [review]

File: CommonClient.py (hunk 3 above)
Comment (Berserker66) [reply CommonClient.py:464]: So consumers of the API can choose a collection. set(set) is basically free anyway.

**Review by black-sliver — APPROVED** [review]
I have not tested this, but the code looks correct to me and the change is good imo.

If you think it needs independent testing, you could wait for qwint, but i think there are no corner cases left if it worked for factorio and lttp.

**Comment by qwint** [comment]
I did just make the switch to the new function on my client and tested sending a couple locations and saw no issues
